zerver/lib/actions.py

```diff
@@ -1931,10 +1931,12 @@ def extract_recipients(
         # We don't complain about empty message recipients here
         return data
 
-    if isinstance(data[0], str):
-        recipients = extract_emails(data)  # type: Union[List[str], List[int]]
-
-    if isinstance(data[0], int):
+    if isinstance(data[0], str) and data[0].isdigit():
+        data = map(int, data)
+        recipients = extract_user_ids(data)  # type: Union[List[str], List[int]]
+    elif isinstance(data[0], str):
+        recipients = extract_emails(data)
+    elif isinstance(data[0], int):
```

Comment:
Can you explain clearly why the Zapier app cannot just convert the strings to user IDs before calling the zulip-js code?  Or if `zulip-js` is the thing converting from integers to strings, why we can't fix that?

Reply:
@aero31aero FYI as another person who knows zulip-js and might be able to help investigate.

Reply:
@timabbott I don't see any mentions of zulip-js in this PR. Is there some broader comtext or some CZO conversation to read?

Reply:
Yeah, the other half of this PR is https://github.com/zulip/zulip-zapier

Reply:
@timabbott: Sure!
* For custom action fields, Zapier requires an argument in the form of an object with key-value pairs such as `raw: label, raw: label`. See [here](https://zapier.com/developer/documentation/v2/reference/#action-fields-custom).
* In our case, the `raw` value needs to be the user ID (since that is what is sent to the API), and the `label` value needs to be the full name of the user (which is what people see in the dynamic drop down in Zapier's UI when deciding who to send the message to). See [the relevant code in zulip/zulip-zapier#6](https://github.com/zulip/zulip-zapier/blob/71ab42ebacfef7eb4c1630d8717e7e906399c764/creates/custom_fields.js#L17).
* Objects in JavaScript by definition [don't support integer keys for objects](https://stackoverflow.com/questions/3633362/is-there-any-way-to-use-a-numeric-type-as-an-object-key), so they are converted to strings by JavaScript, which is what is sent over to `zulip-js` and then to our API.
* I am pretty confident this is not "a misdiagnosed lack of support for sending IDs to zulip-js" because [the relevant function definition](https://github.com/zulip/zulip-js/blob/master/src/resources/messages.js#L15) in `zulip-js` accepts a simple `params` parameter that can definitely contain the `to` argument with user IDs, there is no constraint per se that needs to be accounted for in order to support user IDs here. I've already tested all of these changes with a deployed Zapier app instance.

To conclude, I think this is more so a JS problem, then a `zulip-js` problem. I'm about 98% sure I'm not wrong about this, but that remaining 2% has tripped me up before haha! I hope this clears it up, thanks! :)

Reply:
@timabbott: I will say though that we could (potentially) move this sort of type checking to `zulip-js` if you'd like, where we check if the incoming list of IDs is actually just a list of strings or not. But, even then, I'm thinking that our API should be able to distinguish between IDs represented as strings vs. IDs as ints. :)